cmd/util/cmd/read-badger/cmd/transactions.go

```diff
@@ -39,3 +54,162 @@ var transactionsCmd = &cobra.Command{
+// However there should be at least one block executed for each height, hence adding that check as well
+var transactionsDuplicateCmd = &cobra.Command{
+	Use:   "transactions-duplicates",
+	Short: "report/fix duplicated transactions",
+	Run: func(cmd *cobra.Command, args []string) {
+		storages, db := InitStorages()
+		defer db.Close()
+
+		headers, ok := storages.Headers.(*badger.Headers)
+		if !ok {
+			log.Error().Msg("unsupported headers type")
+			return
+		}
+
+		// txID -> list of blocks
+		megaMap := map[flow.Identifier]map[flow.Identifier]struct{}{}
+
+		missingHeights := map[uint64]struct{}{}
+		allHeights := map[uint64]struct{}{}
+
+		blockNo := 0
+		totalTxs := 0
+
+		indexExistingEntries := 0
+		indexNewEntries := 0
+
+		hasSystemTx := false
+
+		systemTxID := flow.Identifier{}
+
+		batch := badger.NewBatch(db)
+
+		_, err := headers.FindHeaders(func(header *flow.Header) bool {
+
+			if flagMaxBlockHeight > 0 && header.Height > flagMaxBlockHeight {
+				return false
+			}
+
+			if !hasSystemTx {
+
+				systemTx, err := blueprints.SystemChunkTransaction(header.ChainID.Chain())
+				if err != nil {
+					panic(fmt.Sprintf("error while constructing system chunk tx: %s", err))
+				}
+				systemTxID = systemTx.ID()
+
+				hasSystemTx = true
+			}
+
+			blockID := header.ID()
+
+			block, err := storages.Blocks.ByID(blockID)
+			if err != nil {
+				panic(fmt.Sprintf("header %s if here but not block", blockID.String()))
+			}
+
+			missingHeights[header.Height] = struct{}{}
+			allHeights[header.Height] = struct{}{}
+
+			txIndex := uint32(0)
+
+			txs := make([]flow.Identifier, 0)
+
+			for _, guarantee := range block.Payload.Guarantees {
+				lightCollection, err := storages.Collections.LightByID(guarantee.CollectionID)
+				if err != nil {
+
+					log.Warn().Msgf("cannot get light collection %s for block %s (%d), trying full collection: %s", guarantee.CollectionID, blockID.String(), header.Height, err)
+
+					collection, err := storages.Collections.ByID(guarantee.CollectionID)
+
+					if err != nil {
+						panic(fmt.Sprintf("cannot get light and normal collection %s for block %s (%d): %s", guarantee.CollectionID, blockID.String(), header.Height, err))
+					}
+					light := collection.Light()
+					lightCollection = &light
+				}
+
+				txs = append(txs, lightCollection.Transactions...)
+			}
+
+			// add system tx
+			txs = append(txs, systemTxID)
+
+			for _, txID := range txs {
+
+				if _, has := megaMap[txID]; !has {
+					megaMap[txID] = make(map[flow.Identifier]struct{})
+				}
+
+				if _, has := megaMap[txID][blockID]; has {
+					panic(fmt.Sprintf("duplicated tx %s in block %s", txID.String(), blockID.String()))
+				}
+
+				megaMap[txID][blockID] = struct{}{}
+
+				_, err := storages.TransactionResults.ByBlockIDTransactionIndex(blockID, txIndex)
+				if err != nil {
+					if errors.Is(err, storage.ErrNotFound) {
+						transactionResult, err := storages.TransactionResults.ByBlockIDTransactionID(blockID, txID)
```

Comment:
should this report the error?

Reply:
No. It was previously, but turns out not every block received gets executed, so there might not be a tx result for given block and that's OK.
That's why I added a check for heights - there should be at least one tx for every height.

Reply:
Yep, record not found we can skip, but some other problems should bubble up, good catch